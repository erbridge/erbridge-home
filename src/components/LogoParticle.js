import PropTypes from 'prop-types';
import React, { Component, createRef } from 'react';

import './LogoParticle.scss';

class LogoParticle extends Component {
  static propTypes = {
    groupBoundingRect: PropTypes.shape({
      top: PropTypes.number,
      right: PropTypes.number,
    }),
    pointerPosition: PropTypes.shape({
      x: PropTypes.number.isRequired,
      y: PropTypes.number.isRequired,
    }),
    radius: PropTypes.number.isRequired,
    x: PropTypes.number.isRequired,
    y: PropTypes.number.isRequired,
  };

  constructor(props) {
    super(props);

    this.containerRef = createRef();
  }

  // TODO: Smooth the rotation a little to make very fast movement less janky.
  getRotation() {
    const { groupBoundingRect, pointerPosition } = this.props;

    if (!groupBoundingRect && !pointerPosition) {
      return 0;
    }

    if (!this.containerRef.current) {
      return 0;
    }

    const {
      x,
      y,
      width,
      height,
    } = this.containerRef.current.getBoundingClientRect();

    const midPointPosition = { x: x + width / 2, y: y + height / 2 };

    let pointerX = 0;
    let pointerY = 0;

    if (pointerPosition) {
      pointerX = pointerPosition.x || 0;
      pointerY = pointerPosition.y || 0;
    } else if (groupBoundingRect) {
      // Magic numbers estimated from the source svg.
      pointerX = (((groupBoundingRect.right || 0) + 15) * width) / 7.2;
      pointerY = (((groupBoundingRect.top || 0) + 5) * height) / 7.2;
    }

    return (
      (Math.atan2(
        pointerY - midPointPosition.y,
        pointerX - midPointPosition.x,
      ) *
        180) /
      Math.PI
    );
  }

  render() {
    const { radius, x, y } = this.props;
    const rotation = this.getRotation();
    const diameter = 2 * radius;

    return (
      <g
        ref={this.containerRef}
        className="LogoParticle"
        transform={`rotate(${rotation}, ${x}, ${y})`}
      >
        <path
          className="LogoParticle__top-semicircle"
          d={`M${x},${y - radius} a1,1 0 0,0 0,${diameter}`}
        />
        <path
          className="LogoParticle__bottom-semicircle"
          d={`M${x},${y + radius} a1,1 0 0,0 0,${-diameter}`}
        />
      </g>
    );
  }
}

export default LogoParticle;
