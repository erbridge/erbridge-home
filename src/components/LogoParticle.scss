@import '../styles/_colours.scss';
@import '../styles/_transition-timings.scss';

.LogoParticle__top-semicircle,
.LogoParticle__bottom-semicircle {
  transition: fill $hover-transition-duration;
}

.LogoParticle__top-semicircle {
  fill: $light-blue;
}

.LogoParticle__bottom-semicircle {
  fill: $dark-blue;
}

a {
  &:active,
  &:focus,
  &:hover {
    .LogoParticle__top-semicircle {
      fill: $red;
    }

    .LogoParticle__bottom-semicircle {
      fill: $dark-red;
    }
  }
}
