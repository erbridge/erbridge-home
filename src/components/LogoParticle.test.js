import { shallow } from 'enzyme';
import React from 'react';

import LogoParticle from './LogoParticle';

it('renders correctly', () => {
  expect(shallow(<LogoParticle radius={1} x={0} y={0} />)).toMatchSnapshot();
});
