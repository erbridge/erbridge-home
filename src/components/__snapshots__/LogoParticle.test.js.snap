// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<g
  className="LogoParticle"
  transform="rotate(0, 0, 0)"
>
  <path
    className="LogoParticle__top-semicircle"
    d="M0,-1 a1,1 0 0,0 0,2"
  />
  <path
    className="LogoParticle__bottom-semicircle"
    d="M0,1 a1,1 0 0,0 0,-2"
  />
</g>
`;
